StructFields.pm

```diff
@@ -677,7 +677,13 @@ sub emit_struct_fields($$;%) {
         } 'fields-' . $fields_group;
 
         # Needed for unions with fields with non-default ctors (e.g. bitfields)
-        emit "$name(){memset(this, 0, sizeof($name));}";
+        emit_block {
+            local $in_union = 1;
```

Comment:
is this variable unused?

Reply:
No, it's used to tell `render_field_init` that it's rendering fields from a union, and is necessary here to get the behavior we want.

Remember that in Perl, `local` variables are dynamically, not lexically, scoped (see [What is the difference between my and local in Perl?](https://stackoverflow.com/questions/129607/what-is-the-difference-between-my-and-local-in-perl).

Reply:
they're actually pushed into the current block's dynamics symbol table, which _remains visible during a subroutine call_, so any subroutine called will also see them (unless shadowed). lexically scoped `my` locals, however, are only visible _within the lexical block_ (and any enclosed blocks) in which they're created.

it's not so much that they get "restored" but rather that defining a symbol within a block shadows any definition that might exist in a block further up. when the block ends, its symbol tables are disposed (each block effectively has two, one for dynamic scope and one for lexical scope), and any prior definition that was being shadowed is thereby unshadowed and thus once again visible. every block carries with it a pointer to the lexical context in which it was defined (a closure), and the current dynamic symbol context is used to initialize a new dynamic symbol context whenever execution enters a new block.

Reply:
thank for the explanation. my perl fu is low